sqlx-core/src/postgres/listen.rs

```diff
@@ -0,0 +1,188 @@
+
+    /// Build a LISTEN query based on the given channel input.
+    fn build_listen_query(channel: &impl AsRef<str>) -> String {
+        format!(r#"LISTEN "{}";"#, channel.as_ref())
+    }
+}
+
+impl PgPool {
+    /// Fetch a new connection from the pool and register it as a listener on the specified channel.
+    pub async fn listen(&self, channel: &impl AsRef<str>) -> Result<PgListener<PgPoolConnection>> {
+        let mut conn = self.acquire().await?;
+        let cmd = PgConnection::build_listen_query(channel);
+        let _ = conn.execute(cmd.as_str(), Default::default()).await?;
+        Ok(PgListener::new(conn))
+    }
+
+    /// Fetch a new connection from the pool and register it as a listener on the specified channels.
+    pub async fn listen_all(
+        &self,
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<PgPoolConnection>> {
+        let mut conn = self.acquire().await?;
+        for channel in channels {
+            let cmd = PgConnection::build_listen_query(&channel);
+            let _ = conn.execute(cmd.as_str(), Default::default()).await?;
+        }
+        Ok(PgListener::new(conn))
+    }
+}
+
+impl PgPoolConnection {
+    /// Fetch a new connection from the pool and register it as a listener on the specified channel.
+    pub async fn listen(mut self, channel: &impl AsRef<str>) -> Result<PgListener<Self>> {
+        let cmd = PgConnection::build_listen_query(channel);
+        let _ = self.execute(cmd.as_str(), Default::default()).await?;
+        Ok(PgListener::new(self))
+    }
+
+    /// Fetch a new connection from the pool and register it as a listener on the specified channels.
+    pub async fn listen_all(
+        mut self,
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<Self>> {
+        for channel in channels {
+            let cmd = PgConnection::build_listen_query(&channel);
+            let _ = self.execute(cmd.as_str(), Default::default()).await?;
+        }
+        Ok(PgListener::new(self))
+    }
+}
+
+/// A stream of async database notifications.
+///
+/// Notifications will always correspond to the channel(s) specified this object is created.
+pub struct PgListener<C>(C);
+
+impl<C> PgListener<C> {
+    /// Construct a new instance.
+    pub(self) fn new(conn: C) -> Self {
+        Self(conn)
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: AsMut<PgConnection>,
+{
+    /// Get the next async notification from the database.
+    pub async fn next(&mut self) -> Result<NotifyMessage> {
+        loop {
+            match self.0.as_mut().receive().await? {
+                Some(Message::NotificationResponse(notification)) => return Ok(notification.into()),
+                // TODO: verify with team if this is correct. Looks like the connection being closed will cause an error
+                // to propagate up from `recevie`, but it would be good to verify with team.
+                Some(_) | None => continue,
```

Comment:
@mehcode && @abonander just wanted to verify this with you guys. Thoughts?

Reply:
`None` means the connection returned `ReadyForQuery`, so it really just means there's no more messages since the last request. That's probably the right call here but I also want to say that any unexpected messages _should_ be a logic error and maybe shouldn't just get eaten, especially if we're still allowing use of the connection.

Reply:
@abonander do you think I should just raise a protocol error here & then close the connection if we receive something other than a None or a NotificationResponse here?

Reply:
Raising a protocol error, will push the changeset shortly for additional review.